src/Bicep.Core/TypeSystem/DeployTimeConstantContainerVisitor.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class DeployTimeConstantContainerVisitor : SyntaxVisitor
+    {
+        private readonly SemanticModel semanticModel;
+
+        private readonly List<SyntaxBase> deployTimeConstantContainers = new();
+
+        public DeployTimeConstantContainerVisitor(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        public static IEnumerable<SyntaxBase> CollectDeployTimeConstantContainers(SemanticModel semanticModel)
+        {
+            var visitor = new DeployTimeConstantContainerVisitor(semanticModel);
+
+            visitor.Visit(semanticModel.Root.Syntax);
+
+            return visitor.deployTimeConstantContainers;
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            if (syntax.TryGetKeyText() is { } propertyName &&
+                this.semanticModel.Binder.GetParent(syntax) is { } objectSyntax &&
+                this.semanticModel.TypeManager.GetDeclaredType(objectSyntax) is ObjectType objectType &&
```

Comment:
Does this mean we'll ignore DiscriminiatedObjectType instances? Using the assigned type (rather than declared) should have resolved discriminated object types -> object types, as long as the DTC flag has been preserved properly.

Reply:
You are right, we should also handle `DiscriminiatedObjectType`, but using the assigned type won't work for properties whose values contain string interpolations or array accesses, as the DTC flag is not propagated in those cases. I've updated the code to process `DiscriminiatedObjectType` instances and extracted it to a method on `ObjectPropertySyntax`.